mono/mini/trace.c

```diff
@@ -143,9 +143,24 @@ mono_trace_enter_method (MonoMethod *method, char *ebp)
 	g_free (fname);
 
 	if (!ebp) {
+
+// https://github.com/mono/mono/issues/8572
+// Build fails on Redhat 6.8 with GCC 4.4.7.
+// Old gcc does not allow pragma diagnostic within function.
+#if defined (__clang__) || (__GNUC__ > 4) || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6)
+#define GCC_DIAGNOSTIC_WITHIN_FUNCTION 1
+#endif
+
+#if GCC_DIAGNOSTIC_WITHIN_FUNCTION
+#pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wframe-address"
+#endif
```

Comment:
Can't we just move it outside the function ?

Reply:
Yes that would also work. Such distance between pragma and its intended line of code isn't great though, and it covers more code than intended. But yes it would work.